core/arch/x86/decode_fast.c

```diff
@@ -1291,32 +1313,47 @@ decode_cti(dcontext_t *dcontext, byte *pc, instr_t *instr)
     instr_set_opcode(instr, OP_UNDECODED);
     IF_X64(instr_set_x86_mode(instr, get_x86_mode(dcontext)));
 
+    byte0 = *(pc + prefixes);
+    byte1 = *(pc + prefixes + 1);
+
     /* we call instr_set_raw_bits on every return from here, not up
      * front, because any instr_set_src, instr_set_dst, or
-     * instr_set_opcode will kill original bits state */
+     * instr_set_opcode will kill original bits state
+     */
 
     /* Fill in SEG_FS and SEG_GS override prefixes, ignore rest for now.
      * We rely on having these set during bb building.
      * FIXME - could be done in decode_sizeof which is already walking these
      * bytes, but would need to complicate its interface and prefixes are
-     * fairly rare to begin with. */
+     * fairly rare to begin with.
+     */
     if (prefixes > 0) {
         for (i = 0; i < prefixes; i++, pc++) {
             switch (*pc) {
             case FS_SEG_OPCODE: instr_set_prefix_flag(instr, PREFIX_SEG_FS); break;
             case GS_SEG_OPCODE: instr_set_prefix_flag(instr, PREFIX_SEG_GS); break;
+            case VEX_2BYTE_PREFIX_OPCODE:
+                /* VEX 2-byte prefix implies 0x0f opcode */
+                byte0 = 0x0f;
+                byte1 = *(pc + prefixes);
+                break;
+            case VEX_3BYTE_PREFIX_OPCODE: {
+                /* VEX 3-byte prefix implies instruction opcodes by setting mm bits */
+                byte vex_mm = (byte)(*(pc + 1) & 0x1f);
+                get_implied_mm_vex_opcode_bytes(pc, prefixes, vex_mm, &byte0, &byte1);
+                break;
```

Comment:
It would be nice if there were some test that showed the prior bug and proved that this fixes it but we don't have good decode_cti tests, and it's not a public function so it's not as simple to add to the api.ir test.  Please state in the commit message how you tested this.

Reply:
Ok. So far only manually debugged and "verified" with dis-udis86-randtest.raw and an own test I wrote with intrinsics.